### PR TITLE
added option to fix torsional angle

### PR DESCRIPTION
**Pull request by vtlim**

### CODE REVIEW

File: SetUp/autoDefine/autoDefine.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def optsParser( opts ):
 #       keep_going    - If true run define for all directories even if define fails 
 #                       for one. 
 #       save_intermed - If true do not remove files generated in setting up control
-def inputBuilder( directories, options, keep_going, save_intermed ): 
+def inputBuilder( directories, options, keep_going, save_intermed, cosmo ):
```
Comment (magee256) [inline SetUp/autoDefine/autoDefine.py]: Cosmo more logically belongs in the options file, better implementation would be to use a command writer that returns True if $cosmo field is present

File: SetUp/autoDefine/commandWriters.py [hunk 2]
```diff
@@ -193,6 +193,35 @@ def assignSym(botSpecs, entries, defInp):
             else:
                defInp.write('sy '+group+'\n')
 
+#fix() specifies part of geometry to restrain
+#within internal coordinate sub-menu.
+#Only type=tors currently supported.
+#
+#Default: $fix omitted - restraint not used
+#
+#KeyFormat: $fix type=[label] atoms=[label1,label2,label3,label4]
```
Comment (magee256) [inline SetUp/autoDefine/commandWriters.py]: To get multiple specifications a better format would be [type]=[1-4,6]. This allows easy support for all specification types

File: SetUp/autoDefine/commandWriters.py [hunk 3]
```diff
@@ -193,6 +193,35 @@ def assignSym(botSpecs, entries, defInp):
...
+def fix(botSpecs, entries, defInp):
+   key='$fix'
+
+   fix=getLine(entries, key)
+   fixBot=getLine(botSpecs, key)
+
+   if fixBot:
```
Comment (magee256) [inline SetUp/autoDefine/commandWriters.py:219]: Should put a warning here if $fix was already defined, seems likely people would want to tack more frozen coords on.

File: SetUp/autoDefine/autoDefine.py [hunk 4]
```diff
@@ -277,16 +278,19 @@ def inputBuilder( directories, options, keep_going, save_intermed ):
...
-#        p.wait()
-#
-#
-#        #Place setup files no longer needed in setup dir
-#        if not os.path.exists('setup'):
-#            os.makedirs('setup')
-#        for i in ['input.xyz','options','templateCosmo.input','cosmoprep.out','def.input','def.out']:
-#            try: os.rename(i,'setup/'+i)
-#            except OSError: pass
+        if cosmo:
```
Comment (magee256) [inline SetUp/autoDefine/autoDefine.py]: This is better than before, but I still don't want this in the code until cosmoprep.input is generated by the script

File: SetUp/autoDefine/autoDefine.py [hunk 5]
```diff
@@ -321,7 +325,10 @@ def inputBuilder( directories, options, keep_going, save_intermed ):
             help='Continue iterating through directories even if define fails.')
     parser.add_argument('-i','--save_intermed', action='store_true',
             help='Do not remove files generated while setting up control')
+    parser.add_argument('-c','--cosmo', default=False, action='store_true',
```
Comment (magee256) [inline SetUp/autoDefine/autoDefine.py]: Don't make the user specify this on the command line, better to centralize calculation specifications

File: SetUp/autoDefine/commandWriters.py [hunk 6]
```diff
@@ -756,6 +785,7 @@ def scf(botSpecs, entries, defInp):
 
       defInp.write('\n')
 
+
```
Comment (magee256) [inline SetUp/autoDefine/commandWriters.py]: ?

**Review by magee256 — CHANGES_REQUESTED** [review]
A good starting point for a $fix menu, but generalizing to more options is simple and should be done before accepting pull. Also try to keep command line options pertinent to the set up script, not the calculation being set up

File: SetUp/autoDefine/commandWriters.py (hunk 2 above)
Comment (vtlim) [reply SetUp/autoDefine/commandWriters.py]: I don't know what you mean by [1-4,6].

File: SetUp/autoDefine/commandWriters.py (hunk 3 above)
Comment (vtlim) [reply SetUp/autoDefine/commandWriters.py:219]: Again, can you clarify?

**Comment by vtlim** [comment]
I'll work on the cosmo stuff, but if you want to generalize fix, you can do that part.

File: SetUp/autoDefine/commandWriters.py (hunk 3 above)
Comment (magee256) [reply SetUp/autoDefine/commandWriters.py:219]: I can do this. I basically just mean insert the lines: 
```
if fix:
     print('Warning: frozen coordinates specified in top level options file will be overridden')
```

After `if fixBot:`

File: SetUp/autoDefine/commandWriters.py (hunk 2 above)
Comment (magee256) [reply SetUp/autoDefine/commandWriters.py]: That's my shorthand for a list of integers, it's meant to show ranges and comma separated lists are allowed. Maybe not that clear without context.

**Comment by vtlim** [comment]
@magee256, cosmo sections have been updated.

**Pull request merged**